base/compiler/typeinfer.jl

```diff
@@ -850,16 +857,18 @@ function typeinf_edge(interp::AbstractInterpreter, method::Method, @nospecialize
         end
         typeinf(interp, frame)
         update_valid_age!(caller, frame.valid_worlds)
-        edge = is_inferred(frame) ? mi : nothing
-        return EdgeCallResult(frame.bestguess, edge, frame.ipo_effects) # effects are adjusted already within `finish`
+        isinferred = is_inferred(frame)
+        edge = isinferred ? mi : nothing
+        effects = isinferred ? frame.ipo_effects : adjust_effects(Effects(), method) # effects are adjusted already within `finish` for ipo_effects
```

Comment:
I came up with this fix for `typejoin` accuracy for external `AbstractInterpreter` with overlay table:
```suggestion
        if isinferred
            effects = frame.ipo_effects # effects are adjusted already within `finish`
        else
            # if this method isn't overlayed, we don't need to taint `:nonoverlayed`
            effects = is_nonoverlayed(method) ? EFFECTS_UNKNOWN : Effects()
            effects = adjust_effects(effects, method)
        end
```

But I just realized this isn't correct, because we can't guarantee that any overlay methods aren't involved within a recursive call graph..
And now I think this may be more problematic that I thought, because now we basically make external `AbstractInterpreter` with overlay table unable to perform concrete eval on any recursive methods.

I think there are two options:
1. make inference keep track of a set of methods that appear within a call graph and check if there are any overlay methods here, which sounds a bit complicated but at the same time may be useful for the other analyses too
2. move forward #51080 and annotate `:nonoverlayed typejoin`, which seems awkward honestly, because it is obvious that `typejoin` is not overlayed

Reply:
`typejoin` is probably considered to be `:total`, which would possibly already work. The only reason this method is a problem is that it is annotated `:terminates`, since otherwise inference couldn't derive that information itself anyways.